epub33/rs/index.html

```diff
@@ -1852,33 +1867,73 @@ <h4>Scripting Considerations</h4>
 					</li>
 				</ul>
 
-				<p>This specification recommends the following practices for handling untrusted scripts:</p>
-
+				<p>
+					In handling untrusted scripts this specification recommends that Reading Systems establish 
+					a unique <a href="https://url.spec.whatwg.org/#origin">origin</a> [[URL]] allocated to each 
+					<a>EPUB Publication</a> (see <a href="#sec-scripted-content"></a>). Adopting this
+					approach isolates publications from each other, thereby limiting access to cookies, DOM storage, etc. Examples of Web APIs that are tied to the concept of “origin” include Web Storage [[?WEBSTORAGE]] and 
+					IndexedDB [[?INDEXEDDB]], which EPUB content documents can interact with via scripting.
+					This specification recommends the following practices:
+				</p>
 				<ul>
+
+					<li>
+						<p>
+							Although domain isolation is the most effective way to secure EPUB Publications in
+							Web-based Reading Systems, it is not always a realistic option. Web-based Reading
+							Systems must still maintain isolation in such cases, which may require limiting access
+							to scripting APIs that allow client-side data storage and access.
+						</p>
+					</li>
+					<li>
+						<p>
+							Reading Systems that allow users to add/remove publications from a managed library 
+							(a.k.a “bookshelf”) may maintain the publication’s unique origin when the publication is
+							removed and subsequently re-imported into the content library. Conversely, Reading
+							Systems may create a new unique origin for every newly-added publication.
+						</p>
+					</li>
+
+					<li>
+						<p>
+							The unique origin for a given publication (and by extension, for all its content documents)
+							must be preserved across "renders", i.e. when closing/re-opening publications, or when navigating back-and-forth between documents within an opened publication.
```

Comment:
Not sure we need this constraint, and it may not really be true. For instance, in a shared use case it might make sense to assign different origins for each reader of the same local content. I am not saying that is a great design, but I also haven't given enough thought to possible designs to decide if this makes sense. This kind of seems like a recommendation to not break things.

Reply:
@danielweck I am happy to remove this bullet item (@bduga's arguments are compelling); what say you?

Reply:
Well, there are real-world use-cases behind this Reading System conformance requirement, such as:
* EPUBs scripted documents that prompt users to enter their name into an text input field, at an early stage in the reading experience (typically, in the first interactive page after the static cover / intro material), in order to personalise ebook content further into the story narrative. The entered value is preserved in origin-bound LocalStorage.
* EPUBs that store "state" information such as quiz answers, challenge progress, etc. The data structure is saved in origin-bound IndexedDB.
* EPUBs that cache some analytics data, such as "first time" publication or document read event.
* etc.

That being said, there must clearly also be cases where the origin model doesn't matter or is not desirable (due to legacy technical reasons, authoring choices, reading system designs ... or combinations of). So I am not sure how to best reconciliate this, and what kind of testable specification prose can usefully be added / clarified to EPUB v3.x

Reply:
First of all, let us not forget that all these points are part of [§15 Security and Privacy](https://w3c.github.io/epub-specs/epub33/rs/#sec-security-privacy), ie, a _non-normative_ section. In other words, none of these remarks are formal RS conformance requirements, but more a matter of guidance, good practices, that implementers should consider in relations to security and privacy. Hence the usage of “must” in that item as opposed to _“MUST”_.

With that in mind, would it be agreeable to all if we keep that item but replace “must” by “should”?

Reply:
I am fine with replacing must with should, or even "when appropriate". Or even list the specific cases @danielweck mentioned (but that seems harder and longer). Though, looking at that list,  *none* of those reasons seem to be related to security and privacy, except perhaps making things less private. Looking more closely, I don't understand the security and privacy implications of the 3rd and fourth bullets at all. These seem like general tips and tricks for scripting support.

Reply:
> security and privacy

Well, content creators (HTML/JS/CSS developers) expect LocalStorage / IndexedDB to be isolated on a same-origin basis. When a reading system does not ensure distinct origins for different publications, data leaks in such a way that a publication script can (intentionally or inadvertantly) read and write data created in the context of another publication.

This is both a privacy and security concern, wouldn't you agree?

Reply:
I do agree that data leaks are related to privacy and security, but the bullet items I mention don't talk about leaks. The third bullet discusses preserving origins across reading sessions, and the fourth discusses preserving data across reading sessions. While both of those might be important for a general scripting guidance chapter, they don't seem appropriate to a security and privacy one.

Reply:
@bduga @danielweck 

Looking at the list in §15.1:

- The first item may be removed (or moved elsewhere), see https://github.com/w3c/epub-specs/pull/1671#discussion_r633783893

- I tend to agree with @bduga that item 3 ("The unique origin for a given publication...") and item 4 ("Reading Systems must ensure that the data tied to") are not really privacy/security issue. The question is whether they are indeed valuable notes, however, that we want to keep. Some sort of a "usability" notes for RS implementers. Not being a RS implementer, I cannot judge whether those are useful or whether RS implementers would just shrug seeing those items. But indeed, it should probably be moved... elsewhere, but I do not know where.

Do we, should we plan some sort of RS implementers' guide? @dauwhe @wareid @mattgarrish

Reply:
> Do we, should we plan some sort of RS implementers' guide?

I don't think so. We already have plenty of work. We do not have demand for such a thing from potential implementers.

Reply:
@dauwhe I thought so, but I wanted to be sure :-)